File: src/main/kotlin/com/github/frcsty/frozenjoin/action/util/UtilPlayer.java
package com.github.frcsty.frozenjoin.action.util;

import com.github.frcsty.frozenjoin.action.Color;
import net.md_5.bungee.api.ChatMessageType;
import net.md_5.bungee.api.chat.BaseComponent;
import net.md_5.bungee.api.chat.TextComponent;
import org.bukkit.Material;
import org.bukkit.entity.Player;
import org.bukkit.inventory.ItemStack;
import org.bukkit.inventory.meta.ItemMeta;

import java.lang.reflect.Constructor;
import java.lang.reflect.Field;
import java.lang.reflect.Method;
import java.util.Collection;
import java.util.List;
import java.util.Objects;

import static com.github.frcsty.frozenjoin.action.util.Reflection.*;

public final class UtilPlayer {
    private static Class<?> craftPlayerClass;
    private static Method getHandleMethod;
    private static Class<?> entityPlayerClass;
    private static Field playerConnectionField;
    private static Method sendPacketMethod;
    private static Method chatSerializerMethod;
    private static Constructor<?> packetPlayOutChatConstructor;
    private static Constructor<?> packetPlayOutTitleConstructor;
    private static Object titleEnum;
    private static Object subtitleEnum;
    private static Method sendTitleMethod;
    private static Method spigotMethod;
    private static Method sendMessageMethod;

    static {
        if (Reflection.isV1_8()) {
            craftPlayerClass = getCraftBukkitClass("entity.CraftPlayer");
            getHandleMethod = getMethod(craftPlayerClass, "getHandle");

            entityPlayerClass = getNmsClass("EntityPlayer");
            playerConnectionField = getField(entityPlayerClass, "playerConnection");

            final Class<?> packetClass = getNmsClass("Packet");
            final Class<?> playerConnectionClass = getNmsClass("PlayerConnection");
            sendPacketMethod = getMethod(playerConnectionClass, "sendPacket", packetClass);

            final Class<?> iChatBaseComponentClass = getNmsClass("IChatBaseComponent");
            final Class<?> chatSerializerClass = getNmsClass("IChatBaseComponent$ChatSerializer");
            chatSerializerMethod = getMethod(chatSerializerClass, "a", String.class);

            final Class<?> packetPlayOutChatClass = getNmsClass("PacketPlayOutChat");
            packetPlayOutChatConstructor = getConstructor(packetPlayOutChatClass, iChatBaseComponentClass, byte.class);

            final Class<?> packetPlayOutTitleClass = getNmsClass("PacketPlayOutTitle");
            final Class<?> enumTitleActionClass = getNmsClass("PacketPlayOutTitle$EnumTitleAction");

            try {
                titleEnum = getField(enumTitleActionClass, "TITLE").get(null);
                subtitleEnum = getField(enumTitleActionClass, "SUBTITLE").get(null);
            } catch (IllegalAccessException ignored) {
            }

            packetPlayOutTitleConstructor = getConstructor(packetPlayOutTitleClass, enumTitleActionClass, iChatBaseComponentClass, int.class, int.class, int.class);
        } else {
            final Class<?> playerClass = getBukkitClass("entity.Player");
            sendTitleMethod = getMethod(playerClass, "sendTitle", String.class, String.class, int.class, int.class, int.class);

            spigotMethod = getMethod(playerClass, "spigot");
            final Class<?> spigotClass = getBukkitClass("entity.Player$Spigot");

            sendMessageMethod = getMethod(spigotClass, "sendMessage", ChatMessageType.class, BaseComponent.class);
        }
    }

    public static void sendActionbar(final Collection<? extends Player> players, final String msg) {
        if (Reflection.isV1_8()) {
            try {
                final Object iChatBaseComponent = createIChatBaseComponent(msg);
                final Object packetPlayOutChat = packetPlayOutChatConstructor.newInstance(iChatBaseComponent, (byte) 2);

                players.forEach(player -> sendPacket(player, packetPlayOutChat));
            } catch (final ReflectiveOperationException ex) {
                ex.printStackTrace();
            }
        } else {
            try {
                final TextComponent component = new TextComponent(msg);

                for (final Player player : players) {
                    final Object playerSpigot = spigotMethod.invoke(player);
                    sendMessageMethod.invoke(playerSpigot, ChatMessageType.ACTION_BAR, component);
                }
            } catch (final ReflectiveOperationException ex) {
                ex.printStackTrace();
            }
        }
    }

    public static void sendTitle(final Collection<? extends Player> players, final String title, final String subtitle, final int fadeIn, final int stay, final int fadeOut) {
        if (Reflection.isV1_8()) {
            try {
                final Object titleBaseComponent = createIChatBaseComponent(title);
                final Object titlePacket = packetPlayOutTitleConstructor.newInstance(titleEnum, titleBaseComponent, fadeIn, stay, fadeOut);
                Object subtitlePacket = null;

                if (subtitle != null) {
                    final Object subtitleBaseComponent = createIChatBaseComponent(subtitle);
                    subtitlePacket = packetPlayOutTitleConstructor.newInstance(subtitleEnum, subtitleBaseComponent, fadeIn, stay, fadeOut);
                }

                for (final Player player : players) {
                    sendPacket(player, titlePacket);

                    if (subtitlePacket != null) {
                        sendPacket(player, subtitlePacket);
                    }
                }
            } catch (final ReflectiveOperationException ex) {
                ex.printStackTrace();
            }
        } else {
            try {
                for (final Player player : players) {
                    sendTitleMethod.invoke(player, title, subtitle, fadeIn, stay, fadeOut);
                }
            } catch (final ReflectiveOperationException ex) {
                ex.printStackTrace();
            }
        }
    }

    private static Object createIChatBaseComponent(final String text) throws ReflectiveOperationException {
        return chatSerializerMethod.invoke(null, "{\"text\":\"" + text + "\"}");
    }

    private static void sendPacket(final Player player, final Object packet) {
        try {
            final Object playerConnection = playerConnectionField.get(
                    entityPlayerClass.cast(
                            getHandleMethod.invoke(
                                    craftPlayerClass.cast(player)
                            )
                    )
            );

            sendPacketMethod.invoke(playerConnection, packet);
        } catch (final ReflectiveOperationException ex) {
            ex.printStackTrace();
        }
    }

    public static void setPlayerItem(final Player player, final String itemString, final String display, final List<String> lore, final int amount, final int slot) {
        final String[] vars = itemString.split(",");
        final boolean useData = vars.length == 2;
        final Material material = useData ? Material.getMaterial(vars[1]) : Material.getMaterial(itemString);
        final int data = useData ? Integer.valueOf(vars[1]) : 0;

        ItemStack item;
        if (useData) {
            item = new ItemStack(Objects.requireNonNull(material), amount, (short) data);
        } else {
            item = new ItemStack(Objects.requireNonNull(material), amount);
        }

        final ItemMeta meta = item.getItemMeta();
        if (display != null) {
            meta.setDisplayName(Color.colorize(display));
        }
        if (!lore.isEmpty()) {
            meta.setLore(Color.colorize(lore));
        }

        item.setItemMeta(meta);

        if (player.getInventory().getItem(slot) == null) {
            player.getInventory().setItem(slot, item);
        } else {
            player.getInventory().addItem(item);
        }
    }
}

File: src/main/kotlin/com/github/frcsty/frozenjoin/action/util/UtilPlayer.kt
package com.github.frcsty.frozenjoin.action.util

import com.github.frcsty.frozenjoin.extension.color
import net.md_5.bungee.api.ChatMessageType
import net.md_5.bungee.api.chat.BaseComponent
import net.md_5.bungee.api.chat.TextComponent
import org.bukkit.Material
import org.bukkit.entity.Player
import org.bukkit.inventory.ItemStack
import java.lang.reflect.Constructor
import java.lang.reflect.Field
import java.lang.reflect.Method

object UtilPlayer {
    private var craftPlayerClass: Class<*>? = null
    private var getHandleMethod: Method? = null
    private var entityPlayerClass: Class<*>? = null
    private var playerConnectionField: Field? = null
    private var sendPacketMethod: Method? = null
    private var chatSerializerMethod: Method? = null
    private var packetPlayOutChatConstructor: Constructor<*>? = null
    private var packetPlayOutTitleConstructor: Constructor<*>? = null
    private var titleEnum: Any? = null
    private var subtitleEnum: Any? = null
    private var sendTitleMethod: Method? = null
    private var spigotMethod: Method? = null
    private var sendMessageMethod: Method? = null

    fun sendActionbar(players: Collection<Player>, msg: String) {
        if (Reflection.isV1_8) {
            try {
                val iChatBaseComponent = createIChatBaseComponent(msg)
                val packetPlayOutChat = packetPlayOutChatConstructor!!.newInstance(iChatBaseComponent, 2.toByte())
                players.forEach { player: Player -> sendPacket(player, packetPlayOutChat) }
            } catch (ex: ReflectiveOperationException) {
                ex.printStackTrace()
            }
        } else {
            try {
                val component = TextComponent(msg)
                for (player in players) {
                    val playerSpigot = spigotMethod!!.invoke(player)
                    sendMessageMethod!!.invoke(playerSpigot, ChatMessageType.ACTION_BAR, component)
                }
            } catch (ex: ReflectiveOperationException) {
                ex.printStackTrace()
            }
        }
    }

    fun sendTitle(players: Collection<Player>, title: String, subtitle: String?, fadeIn: Int, stay: Int, fadeOut: Int) {
        if (Reflection.isV1_8) {
            try {
                val titleBaseComponent = createIChatBaseComponent(title)
                val titlePacket = packetPlayOutTitleConstructor!!.newInstance(titleEnum, titleBaseComponent, fadeIn, stay, fadeOut)
                var subtitlePacket: Any? = null
                if (subtitle != null) {
                    val subtitleBaseComponent = createIChatBaseComponent(subtitle)
                    subtitlePacket = packetPlayOutTitleConstructor!!.newInstance(subtitleEnum, subtitleBaseComponent, fadeIn, stay, fadeOut)
                }
                for (player in players) {
                    sendPacket(player, titlePacket)
                    subtitlePacket?.let { sendPacket(player, it) }
                }
            } catch (ex: ReflectiveOperationException) {
                ex.printStackTrace()
            }
        } else {
            try {
                for (player in players) {
                    sendTitleMethod!!.invoke(player, title, subtitle, fadeIn, stay, fadeOut)
                }
            } catch (ex: ReflectiveOperationException) {
                ex.printStackTrace()
            }
        }
    }

    @Throws(ReflectiveOperationException::class)
    private fun createIChatBaseComponent(text: String): Any {
        return chatSerializerMethod!!.invoke(null, "{\"text\":\"$text\"}")
    }

    private fun sendPacket(player: Player, packet: Any) {
        try {
            val playerConnection = playerConnectionField!![entityPlayerClass!!.cast(
                    getHandleMethod!!.invoke(
                            craftPlayerClass!!.cast(player)
                    )
            )]
            sendPacketMethod!!.invoke(playerConnection, packet)
        } catch (ex: ReflectiveOperationException) {
            ex.printStackTrace()
        }
    }

    fun setPlayerItem(player: Player, itemString: String, display: String?, lore: List<String?>, amount: Int, slot: Int) {
        val vars = itemString.split(",").toTypedArray()
        val useData = vars.size == 2
        val material = (if (useData) Material.getMaterial(vars[1]) else Material.getMaterial(itemString))
                ?: Material.STONE
        val data = if (useData) Integer.valueOf(vars[1]) else 0
        val item: ItemStack

        item = if (useData) {
            ItemStack(material, amount, data.toShort())
        } else {
            ItemStack(material, amount)
        }
        val meta = item.itemMeta
        if (display != null) {
            meta.setDisplayName(display.color())
        }
        if (lore.isNotEmpty()) {
            meta.lore = lore
        }
        item.itemMeta = meta
        if (player.inventory.getItem(slot) == null) {
            player.inventory.setItem(slot, item)
        } else {
            player.inventory.addItem(item)
        }
    }

    init {
        if (Reflection.isV1_8) {
            craftPlayerClass = Reflection.getCraftBukkitClass("entity.CraftPlayer")
            getHandleMethod = Reflection.getMethod(craftPlayerClass!!, "getHandle")
            entityPlayerClass = Reflection.getNmsClass("EntityPlayer")
            playerConnectionField = Reflection.getField(entityPlayerClass!!, "playerConnection")
            val packetClass = Reflection.getNmsClass("Packet")
            val playerConnectionClass = Reflection.getNmsClass("PlayerConnection")
            sendPacketMethod = Reflection.getMethod(playerConnectionClass, "sendPacket", packetClass)
            val iChatBaseComponentClass = Reflection.getNmsClass("IChatBaseComponent")
            val chatSerializerClass = Reflection.getNmsClass("IChatBaseComponent\$ChatSerializer")
            chatSerializerMethod = Reflection.getMethod(chatSerializerClass, "a", String::class.java)
            val packetPlayOutChatClass = Reflection.getNmsClass("PacketPlayOutChat")
            packetPlayOutChatConstructor = Reflection.getConstructor(packetPlayOutChatClass, iChatBaseComponentClass, Byte::class.javaPrimitiveType)
            val packetPlayOutTitleClass = Reflection.getNmsClass("PacketPlayOutTitle")
            val enumTitleActionClass = Reflection.getNmsClass("PacketPlayOutTitle\$EnumTitleAction")
            try {
                titleEnum = Reflection.getField(enumTitleActionClass, "TITLE")[null]
                subtitleEnum = Reflection.getField(enumTitleActionClass, "SUBTITLE")[null]
            } catch (ignored: IllegalAccessException) {
            }
            packetPlayOutTitleConstructor = Reflection.getConstructor(packetPlayOutTitleClass, enumTitleActionClass, iChatBaseComponentClass, Int::class.javaPrimitiveType, Int::class.javaPrimitiveType, Int::class.javaPrimitiveType)
        } else {
            val playerClass = Reflection.getBukkitClass("entity.Player")
            sendTitleMethod = Reflection.getMethod(playerClass, "sendTitle", String::class.java, String::class.java, Int::class.javaPrimitiveType, Int::class.javaPrimitiveType, Int::class.javaPrimitiveType)
            spigotMethod = Reflection.getMethod(playerClass, "spigot")
            val spigotClass = Reflection.getBukkitClass("entity.Player\$Spigot")
            sendMessageMethod = Reflection.getMethod(spigotClass, "sendMessage", ChatMessageType::class.java, BaseComponent::class.java)
        }
    }
}